// o-stream/lib/main.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var ObjectTransform_1 = require("./ObjectTransform");
exports.ObjectTransform = ObjectTransform_1.ObjectTransform;
var stream_1 = require("stream");
exports.Writable = stream_1.Writable;
exports.Readable = stream_1.Readable;
exports.Transform = stream_1.Transform;
__export(require("./EnteredArgs"));
__export(require("./EndedArgs"));
var Default = (function () {
    function Default() {
    }
    Default.transform = function (args) {
        return new ObjectTransform_1.ObjectTransform(args);
    };
    Default.fromArray = function (array) {
        var i = 0;
        var stream = new ReadableObjectsStream(function () {
            if (i == array.length) {
                return null;
            }
            return array[i++];
        });
        return stream;
    };
    return Default;
}());
exports.default = Default;
var ReadableObjectsStream = (function (_super) {
    __extends(ReadableObjectsStream, _super);
    function ReadableObjectsStream(readObject) {
        var _this = _super.call(this, { objectMode: true }) || this;
        _this.readObject = readObject;
        return _this;
    }
    ReadableObjectsStream.prototype._read = function () {
        var shouldPushMore = true;
        while (shouldPushMore) {
            shouldPushMore = this.push(this.readObject());
        }
    };
    return ReadableObjectsStream;
}(stream_1.Readable));
